### DISCUSSION TITLE
Safety of accessing freed GCHandle

### DISCUSSION THREAD

**Discussion by acaly** [dotnet/runtime#46502] [Q&A] [answered]
When using `GCHandle`, I noticed that the `Free` method atomically sets the handle to 0, which causes further use of `Target` property to throw. However, if I have another copy of the same handle (since it's a value type), I can still access that handle, which seems to always return `null` from the `Target` property. I would like to know whether this is reliable, i.e., whether accessing the `Target` property of `Free`d handle is safe and always returns null.

I expect the answer to be the same for all kinds of handles, but I am mainly using weak handles.

Thanks!

**Comment by PathogenDavid** [comment, marked as answer]
> I would like to know whether this is reliable, i.e., whether accessing the `Target` property of `Free`d handle is safe and always returns null.

It is not safe or reliable. Internally GCHandle is essentially an index into a table of references. When you free a handle, that entry of the table is deallocated and set to null.

When you're accessing `Target` of the dead handle, you're accessing that deallocated row of the table. If that row gets reused, it will no longer be null and will point to something else entirely.

You can observe this by allocating another handle immediately after freeing one:

```csharp
using System;
using System.Runtime.InteropServices;

string a = "StringA";
string b = "StringB";
GCHandle handle = GCHandle.Alloc(a);
GCHandle clone = handle;
Console.WriteLine(clone.Target ?? "Null");
handle.Free();
Console.WriteLine(clone.Target ?? "Null");
GCHandle anotherHandle = GCHandle.Alloc(b);
Console.WriteLine(clone.Target ?? "Null");
```

Running this program will result in:

```
StringA
Null
StringB
```

**Reply by acaly** [reply]
Thank you for the answer! 

Actually I am fine with the handle pointing to other places. I have a logic to confirm the `Target` is correct, but if it's stored like a table and the runtime shrinks the table in a large program I might get access violation.

Maybe I should consider making a pool to reuse the `GCHandle`s instead of `Free`ing them. 😕

**Comment by huoyaoyuan** [comment]
If you don't have any strong requirement, using `WeakReference` or `fixed` should be quite better than `GCHandle`.

**Reply by acaly** [reply]
Thanks for the suggestion. The reason I am using GCHandle is because it's an unmanaged type and is fast. I am also keeping those handles for long so fixed might not be a good choice.
